src/solver/conv/conv_direct_naive_conv.cpp

```diff
@@ -361,11 +361,27 @@ GetConv2DFWDSolution(const ExecutionContext& ctx, const ::miopen::conv::ProblemD
     size_t grid_size  = 1;
     if(problem.IsLayoutDefault())
     {
-        grid_size = static_cast<size_t>(n) * k;
+        size_t all_workload = static_cast<size_t>(n) * k;
+        if(all_workload <= block_size)
+        {
+            grid_size = all_workload;
+        }
+        else
+        {
+            grid_size = (all_workload + block_size - 1) / block_size;
+        }
     }
     else if(problem.IsLayoutNHWC())
     {
-        grid_size = static_cast<size_t>(group) * n * ho;
+        size_t all_workload = static_cast<size_t>(group) * n * ho;
+        if(all_workload <= block_size)
+        {
+            grid_size = all_workload;
+        }
+        else
+        {
+            grid_size = (all_workload + block_size - 1) / block_size;
+        }
```

Comment:
I don't feel like the problem is solved here, actually I see a few more problems.
Now it divides total workload by 256 - technically speaking it's just 256 times further from now. Quite far away, but still there.
And since it is dividing total workload by 256, we have 256 times underloaded GPU. Can be a huge performance drop for a wide range of legit tensor sizes, and even it's a naive algorithm, we are using it everywhere in the tests to compute reference data.
The last concern is the kernel itself - it should be aware about that fact that the number of groups can be capped, and it should contain extra loop to handle it.

Reply:
I will implemented the kernel itself to handle the capped number of groups.

Reply:
I'm not sure that a new kernel should be implemented, or the old one can be changed, or even the old one has already got this support and we should change anything - firstly it should be checked.

Underloaded GPU problem should be fixed too.
Let's imagine - `all_workload` is 256 and we have a grid size of 256; when it is 257, the grid size suddenly becomes 2.
We have more work but fewer workers.

Reply:
Maybe we don't need to modify the kernel. We could loop over the same kernel, adjusting the chunk size and buffer offsets as needed. This would handle the limitation of uint32_t in  [hipExtModuleLaunchKernel](https://github.com/ROCm/MIOpen/blob/develop/src/hipoc/hipoc_kernel.cpp#L117) which currently overflows when we pass a grid size like [gridX((589824 *256) *256 )](https://github.com/ROCm/MIOpen/blob/develop/src/solver/conv/conv_direct_naive_conv.cpp#L379).

Reply:
I was looking to see how we handle this issue in other locations (since it seems like it would be a global constraint). 
 
Looks like the [batched_transpose solver](https://github.com/ROCm/MIOpen/blob/9b4cf56919c85aae26a3eb9206b16dd9506c171b/src/hip/batched_transpose_sol.cpp#L340-L397) also has a version of this issue (and seems somewhat likely we have this issue throughout MIOpen).
 
For HIP this is a general constraint across any kernel launch I think:

```
What are the maximum limits of kernel launch parameters?

Product of block.x, block.y, and block.z should be less than 1024. Please note, HIP does not support kernel launch with total work items defined in dimension with size gridDim x blockDim >= 2^32, so gridDim.x * blockDim.x, gridDim.y * blockDim.y and gridDim.z * blockDim.z are always less than 2^32.
```

I think we might need to come up with a general solution for this, and make sure it's implemented broadly.

Reply:
> I was looking to see how we handle this issue in other locations (since it seems like it would be a global constraint).
> 

We don't, kind of. There are some places where the kernel is aware about number of workgroups limit and sometimes the number of workgroups is capped by some value like 4096. That's mostly it.

> I think we might need to come up with a general solution for this, and make sure it's implemented broadly.
>

I'm not sure if thatcan be easily implemented. The main reason is that: the number of workgroups heavily depends on the algorithm and, the most important, on the kernel itself, and sometimes it even comes from heuristics.
Putting some hardlimit in the library will not resolve the problem, and it can even do a bad stuff like previously runtime explicitly failed to launch the kernel, but now it will be silently capped, launched and produced a wrong result, which will be much harded to notice, especially when you run production code without any verifications.